qiskit_dynamics/pulse/string_model_parser/operator_from_string.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+# pylint: disable=invalid-name
+
+"""Generate operators from string."""
+
+import numpy as np
+
+from qiskit import QiskitError
+
+
+def operator_from_string(op_label: str, subsystem_index: int, subsystem_dims: dict) -> np.ndarray:
+    r"""Generates a dense operator acting on a single subsystem, tensoring
+    identities for remaining subsystems.
+
+    The single system operator is specified via a string in ``op_label``,
+    the list of subsystems and their corresponding dimensions are specified in the
+    dictionary ``subsystem_dims``, with system label being the keys specified as ``int``s,
+    and system dimensions the values also specified as ``int``s, and ``subsystem_index``
+    indicates which subsystem the operator specified by ``op_label`` acts on.
+
+    Accepted ``op_labels`` are:
+        - `'X'`: If the target subsystem is two dimensional, the
+          Pauli :math:`X` operator, and if greater than two dimensional, returns
+          :math:`a + a^\dagger`, where :math:`a` and :math:`a^\dagger` are the
+          annihiliation and creation operators, respectively.
+        - `'Y'`: If the target subsystem is two dimensional, the
+          Pauli :math:`Y` operator, and if greater than two dimensional, returns
+          :math:`-i(a - a^\dagger)`, where :math:`a` and :math:`a^\dagger` are the
+          annihiliation and creation operators, respectively.
+        - `'Z'`: If the target subsystem is two dimensional, the
+          Pauli :math:`Z` operator, and if greater than two dimensional, returns
+          :math:`I - 2 * N`, where :math:`N` is the number operator.
+        - `'a'`, `'A'`, or `'Sm'`: If two dimensional, the sigma minus operator, and if greater,
+          generalizes to the operator.
+        - `'C'`, or `'Sp'`: If two dimensional, sigma plus operator, and if greater,
+          generalizes to the creation operator.
+        - `'N'`, or `'O'`: The number operator.
+        - `'I'`: The identity operator.
+
+    Note that the ordering of tensor factors is reversed.
+
+    Args:
+        op_label: The string labelling the single system operator.
+        subsystem_index: Index of the subsystem to apply the operator.
```

Comment:
This subsystem index is just a single `int` indicating the label of the subsystem that the operator being constructed acts on. The `subsystem_dims` argument is a dictionary mapping subsystem labels to dimensions, so that this function knows where and what order to tensor on identities. The subsystem labels in `subsystem_dims` are assumed to be ints by this function.

To address this:
- Made the type hint for `subsystem_labels` more detailed to show that both the keys and labels are ints.
- Changing `subsystem_index` to `subsystem_label` to avoid confusion that "index" might be indexing a list of subsystems (as opposed to specifying a specific label).